include/StateChecker.php

```diff
@@ -0,0 +1,322 @@
+     */
+    public function __construct()
+    {
+        if (!$this->db = DBManagerFactory::getInstance()) {
+            throw new StateCheckerException('DBManagerFactory get instace failure');
+        }
+        if (!($this->db instanceof MysqliManager)) {
+            throw new StateCheckerException('Incompatible DB type, only supported: mysqli');
+        }
+        $this->resetHashes();
+        $this->resetTraces();
+        
+        if (StateCheckerConfig::get('redefineMemoryLimit')) {
+            $this->memoryLimit = ini_get('memory_limit');
+            ini_set('memory_limit', -1);
+        }
+                
+        if (StateCheckerConfig::get('autoRun')) {
+            $this->getStateHash();
+        }
+    }
+    
+    /**
+     * 
+     * @return array traces
+     * @throws StateCheckerException
+     */
+    public function getTraces()
+    {
+        if (StateCheckerConfig::get('saveTraces')) {
+            throw new StateCheckerException('Trace information is not saved, use StateCheckerConfig::get(\'saveTraces\') as true');
+        }
+        return $this->traces;
+    }
+    
+    /**
+     * 
+     */
+    public function __destruct()
+    {
+        if (StateCheckerConfig::get('redefineMemoryLimit')) {
+            ini_set('memory_limit', $this->memoryLimit);
+        }
+    }
+    
+    /**
+     * resetTraces
+     */
+    protected function resetTraces()
+    {
+        $this->traces = [];
+    }
+
+    /**
+     * resetHashes
+     */
+    protected function resetHashes()
+    {
+        $this->hashes = [];
+    }
+    
+    /**
+     * 
+     * @param string $key
+     * @return boolean
+     */
+    protected function isDetailedKey($key)
+    {
+        $detailedKey = preg_match('/\w+\:\:/', $key);
+        return $detailedKey;
+    }
+    
+    /**
+     * 
+     * @param string $hash
+     * @param string $key
+     * @return boolean
+     */
+    protected function checkHash($hash, $key)
+    {
+        $detailedKey = $this->isDetailedKey($key);
+        $needToStore = !$detailedKey || ($detailedKey && StateCheckerConfig::get('storeDetails'));
+        
+        if (!isset($this->hashes[$key])) {
+            if ($needToStore) {
+                $this->hashes[$key] = $hash;
+            }
+        }
+        if ($needToStore) {
+            $match = $this->hashes[$key] == $hash;
+        } else {
+            $match = true;
+        }
+        return $match;
+    }
+    
+    /**
+     * 
+     * @param mixed $data should be serializable
+     * @param string $key
+     * @return string
+     * @throws StateCheckerException
+     */
+    protected function getHash($data, $key)
+    {
+        if (!$serialized = serialize($data)) {
+            throw new StateCheckerException('Serialize object failure');
+        }
+        $hash = md5($serialized);
+        
+        if (!$this->checkHash($hash, $key)) {
+            throw new StateCheckerException('Hash doesn\'t match at key "' . $key . '".');
```

Comment:
Just an advice. You can use sprintf for a case like this [when a variable in the middle of a string]. I think it makes easier and nicer.

Reply:
Thank, yes, I know, also the language translation handled one of my other PR which is not merged yet.